File: typescript/tests/giftcard/application/integration/booking/handlers/creditFromBooking.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import {
  bookingCompletedIntegrationEvent,
  bookingRefusedIntegrationEvent,
  bookingRejectedIntegrationEvent,
} from '@/booking/integration/giftcard/bookingResultIntegrationEvent.js';
import { CreditFromBooking } from '@/giftcard/application/integration/booking/handlers/creditFromBooking.js';
import { GiftCardCrediting } from '@/giftcard/application/usecases/giftCardCrediting.js';
import { BookingResult } from '@/giftcard/application/integration/booking/adapter/bookingResult.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { SqliteGiftCardRepository } from '@/giftcard/infrastructure/sqliteGiftCardRepository.js';
import { DatabaseSetup } from '../../../../../testsupport/databaseSetup.js';
import { GiftCardAggregateFactory } from '../../../../../testsupport/giftcard/aggregateFactory.js';

describe('CreditFromBooking', () => {
  let repository: SqliteGiftCardRepository;
  let creditFromBooking: CreditFromBooking;

  beforeAll(() => {
    const database = DatabaseSetup.initializeInMemoryDb('giftcard');
    repository = new SqliteGiftCardRepository(database);
    const bookingResult = new BookingResult();
    const useCase = new GiftCardCrediting(repository);
    creditFromBooking = new CreditFromBooking(bookingResult, useCase);
  });

  describe('construction', () => {
    it('rejects null parameters', () => {
      const bookingResult = new BookingResult();
      const useCase = new GiftCardCrediting(repository);

      expect(() => new CreditFromBooking(null as unknown as BookingResult, useCase)).toThrow();
      expect(() => new CreditFromBooking(bookingResult, null as unknown as GiftCardCrediting)).toThrow();
    });
  });

  describe('booking results handling', () => {
    it('on completed should increase balance', () => {
      const giftCard = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const credit = new Money(14);

      creditFromBooking.handle(
        bookingCompletedIntegrationEvent(giftCard.id().value.value, credit)
      );

      const updated = repository.findById(giftCard.id());
      expect(updated).not.toBeNull();
      expect(updated!.balance()).toEqual(credit);
    });

    it('on refused should increase balance', () => {
      const giftCard = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const credit = new Money(9.99);

      creditFromBooking.handle(
        bookingRefusedIntegrationEvent(giftCard.id().value.value, credit)
      );

      const after = repository.findById(giftCard.id());
      expect(after).not.toBeNull();
      expect(after!.balance()).toEqual(credit);
    });

    it('on rejected should do nothing', () => {
      const giftCard = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const amount = new Money(5);

      creditFromBooking.handle(
        bookingRejectedIntegrationEvent(giftCard.id().value.value, amount)
      );

      const persisted = repository.findById(giftCard.id());
      expect(persisted).not.toBeNull();
      expect(persisted!.balance()).toEqual(Money.zero());
    });

    it('should fail if card does not exist', () => {
      const nonExisting = generateId((value) => new GiftCardId(value));
      const amount = new Money(5);

      expect(() =>
        creditFromBooking.handle(
          bookingCompletedIntegrationEvent(nonExisting.value.value, amount)
        )
      ).toThrow();
    });
  });
});

File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/integration/booking/handlers/CreditFromBookingTest.java
package it.giannibombelli.wsc2026.giftcard.application.integration.booking.handlers;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.integration.giftcard.BookingResultIntegrationEvent;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.application.integration.booking.adapter.BookingResult;
import it.giannibombelli.wsc2026.giftcard.application.usecases.GiftCardCrediting;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import it.giannibombelli.wsc2026.giftcard.infrastructure.SqliteGiftCardRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.Optional;

import static it.giannibombelli.wsc2026.testsupport.AggregateFactory.createBooking;
import static it.giannibombelli.wsc2026.testsupport.AggregateFactory.getSavedGiftCard;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class CreditFromBookingTest {

    private GiftCardRepository repository;
    private CreditFromBooking creditFromBooking;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("giftcard");
        repository = new SqliteGiftCardRepository(dataSource);

        BookingResult bookingResult = new BookingResult();
        GiftCardCrediting useCase = new GiftCardCrediting(repository);
        creditFromBooking = new CreditFromBooking(bookingResult, useCase);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullParameters() {
            BookingResult bookingResult = new BookingResult();
            GiftCardCrediting useCase = new GiftCardCrediting(repository);

            assertThatThrownBy(() -> new CreditFromBooking(null, useCase))
                .isInstanceOf(NullPointerException.class);

            assertThatThrownBy(() -> new CreditFromBooking(bookingResult, null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class BookingResultsHandling {
        @Test
        void onConfirmed_shouldIncreaseBalance() {
            GiftCard giftCard = getSavedGiftCard(repository);
            Booking booking = createBooking();
            Money credit = new Money(new BigDecimal("14.00"));

            creditFromBooking.handle(
                new BookingResultIntegrationEvent.BookingCompletedIntegrationEvent(giftCard.id().value(), credit)
            );

            Optional<GiftCard> updated = repository.findById(giftCard.id());
            assertThat(updated).isPresent();
            assertThat(updated.get().balance()).isEqualTo(credit);
        }

        @Test
        void onRefused_shouldIncreaseBalance() {
            GiftCard giftCard = getSavedGiftCard(repository);
            final Booking booking = createBooking();
            Money credit = new Money(new BigDecimal("9.99"));

            creditFromBooking.handle(
                new BookingResultIntegrationEvent.BookingRefusedIntegrationEvent(giftCard.id().value(), credit)
            );

            GiftCard after = repository.findById(giftCard.id()).orElseThrow();
            assertThat(after.balance()).isEqualTo(credit);
        }

        @Test
        void onRejected_shouldDoNothing() {
            GiftCard giftCard = getSavedGiftCard(repository);
            Booking booking = createBooking();
            Money amount = new Money(new BigDecimal("5.00"));

            creditFromBooking.handle(
                new BookingResultIntegrationEvent.BookingRejectedIntegrationEvent(giftCard.id().value(), amount)
            );

            GiftCard persisted = repository.findById(giftCard.id()).orElseThrow();
            assertThat(persisted.balance()).isEqualTo(Money.zero());
        }

        @Test
        void shouldFailIfCardDoesNotExist() {
            GiftCardId nonExisting = EntityId.generate(GiftCardId::new);
            Booking booking = createBooking();
            Money amount = new Money(new BigDecimal("5.00"));

            assertThatThrownBy(() -> creditFromBooking.handle(
                new BookingResultIntegrationEvent.BookingCompletedIntegrationEvent(nonExisting.value(), amount)
            ))
                .isInstanceOf(IllegalStateException.class);
        }
    }
}
